src/modules/audio/audio.c

```diff
@@ -31,90 +42,192 @@ struct Source {
-    ma_uint64 framesIn = source->sound->read(source->sound, source->offset, chunk, raw);
-    ma_uint64 framesOut = sizeof(aux) / (sizeof(float) * outputChannelCountForSource(source));
-
-    ma_data_converter_process_pcm_frames(source->converter, raw, &framesIn, aux, &framesOut);
+// Return value is number of stereo frames in buffer
+// Note: output is always equal to scratchBuffer1. This saves a little memory but is ugly.
+//       count is always less than or equal to the size of scratchBuffer1
+static int generateSource(Source* source, float* output, uint32_t count) {
+  char *raw; float *aux; // Scratch buffers: Raw generated audio from source; converted to float by converter
+  if (source->spatial) { // In spatial mode, raw and aux are mono and only output is stereo
+    raw = (char *)state.scratchBuffer1;
+    aux = state.scratchBuffer2;
+  } else {               // Otherwise, the data converter will produce stereo and aux=output is stereo
+    raw = (char *)state.scratchBuffer2;
+    aux = output;
+  }
 
-    if (source->spatial) {
-      state.spatializer->apply(source, source->transform, aux, mix, framesOut);
+  bool sourceFinished = false;
+  ma_uint64 framesIn = 0;
+  while (framesIn < count) { // Read from source until raw buffer filled
+    ma_uint64 framesRequested = count - framesIn;
+    // FIXME: Buffer size math will break (crash) if channels > 2
+    ma_uint64 framesRead = source->sound->read(source->sound, source->offset, framesRequested,
+      raw + framesIn * SampleFormatBytesPerFrame(source->sound->channels, source->sound->format));
+    framesIn += framesRead;
+    if (framesRead < framesRequested) {
+      source->offset = 0;
+      if (!source->looping) { // Source has reached its final end
+        sourceFinished = true;
+        break;
+      }
     } else {
-      memcpy(mix, aux, framesOut * SampleFormatBytesPerFrame(OUTPUT_CHANNELS, SAMPLE_F32));
+      source->offset += framesRead;
     }
+  }
+  // 1 channel for spatial, 2 otherwise
+  ma_uint64 framesConverted = framesIn * outputChannelCountForSource(source);
 
-    for (uint32_t i = 0; i < framesOut * OUTPUT_CHANNELS; i++) {
-      output[i] += mix[i] * source->volume;
-    }
+  // We assume framesConverted is not changed by calling this and discard its value
+  ma_data_converter_process_pcm_frames(source->converter, raw, &framesIn, aux, &framesConverted);
 
-    if (framesIn == 0) {
-      source->offset = 0;
-      if (!source->looping) {
-        source->playing = false;
-        return false;
-      }
-    } else {
-      source->offset += framesIn;
+  ma_uint64 framesOut = framesIn;
+
+  if (source->spatial) {
+    // Fixed buffer mode we have to pad buffer with silence if it underran
+    if (state.fixedBuffer && sourceFinished) {
+      memset(aux + framesIn, 0, (count - framesIn)*sizeof(float)); // Note always mono
```

Comment:
I wonder if this ought to use `framesConverted` instead of `framesIn` -- the aux buffer stores converted samples, but this line is scrolling through the aux buffer based on the number of raw samples *read* (framesIn), instead of the number of samples *converted* (framesConverted).  This could lead to accidentally zeroing out valid data or leaving in garbage data if conversion leads to a different sample count.  Do I have that right?  Still squinting at it